File: app/src/main/java/com/ivanovsky/passnotes/presentation/core/ScreenState.java
package com.ivanovsky.passnotes.presentation.core;

public class ScreenState {

	public static ScreenState loading() {
		ScreenState result = new ScreenState();

		result.state = FragmentState.LOADING;

		return result;
	}

	public static ScreenState empty(String message) {
		ScreenState result = new ScreenState();

		result.state = FragmentState.EMPTY;
		result.message = message;

		return result;
	}

	public static ScreenState error(String message) {
		ScreenState result = new ScreenState();

		result.state = FragmentState.ERROR;
		result.message = message;

		return result;
	}

	public static ScreenState data() {
		ScreenState result = new ScreenState();

		result.state = FragmentState.DISPLAYING_DATA;

		return result;
	}

	public static ScreenState dataWithError(String message) {
		ScreenState result = new ScreenState();

		result.state = FragmentState.DISPLAYING_DATA_WITH_ERROR_PANEL;
		result.message = message;

		return result;
	}

	private String message;
	private FragmentState state;

	private ScreenState() {
	}

	public String getMessage() {
		return message;
	}

	public FragmentState getState() {
		return state;
	}
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/presentation/core/ScreenState.kt
package com.ivanovsky.passnotes.presentation.core

import androidx.compose.runtime.Immutable
import com.ivanovsky.passnotes.data.entity.OperationError

@Immutable
data class ScreenState private constructor(
    val type: ScreenStateType,
    val emptyText: String? = null,
    val error: OperationError? = null,
    val errorButtonText: String? = null
) {

    val isDisplayingData: Boolean
        get() = (type == ScreenStateType.DATA || type == ScreenStateType.DATA_WITH_ERROR)

    val isDisplayingErrorPanel: Boolean
        get() = (type == ScreenStateType.DATA_WITH_ERROR)

    val isDisplayingLoading: Boolean
        get() = (type == ScreenStateType.LOADING)

    val isDisplayingEmptyState: Boolean
        get() = (type == ScreenStateType.EMPTY)

    val isNotInitialized: Boolean
        get() = (type == ScreenStateType.NOT_INITIALIZED)

    val isDisplayingError: Boolean
        get() = (type == ScreenStateType.ERROR)

    companion object {

        fun data(): ScreenState =
            ScreenState(ScreenStateType.DATA)

        fun loading(): ScreenState =
            ScreenState(ScreenStateType.LOADING)

        fun empty(emptyText: String?): ScreenState =
            ScreenState(
                ScreenStateType.EMPTY,
                emptyText = emptyText
            )

        fun notInitialized(): ScreenState =
            ScreenState(ScreenStateType.NOT_INITIALIZED)

        fun error(error: OperationError): ScreenState =
            ScreenState(
                type = ScreenStateType.ERROR,
                emptyText = null,
                error = error
            )

        fun dataWithError(error: OperationError): ScreenState =
            ScreenState(
                type = ScreenStateType.DATA_WITH_ERROR,
                emptyText = null,
                error = error
            )

        fun dataWithError(error: OperationError, errorButtonText: String? = null): ScreenState =
            ScreenState(
                type = ScreenStateType.DATA_WITH_ERROR,
                emptyText = null,
                error = error,
                errorButtonText = errorButtonText
            )
    }
}